### PR TITLE
A new option to SHC

### PR DESCRIPTION
**Pull request by manxkim**
This PR implements Ji Hoon Ryoo's spin Hall conductivity using Wannier interpolation, tested and publised(Phys. Rev. B 99, 235113). It is slightly different from Junfeng Qiao's (Phys. Rev. B 98, 214402), which uses some approximations.

To calculate SHC with Ryoo/Qiao's method, add the parameter "shc_method=ryoo" or "qiao" in postw90.x input description. It can be used together with other shc parameters by Qiao. Also, it requires .sHu and .sIu files generated from pw2wannier90.x of quantum espresso, implemented by Ryoo and me.

Continued to the conversation https://github.com/wannier-developers/wannier90/pull/350:

1. The parameter shc_ryoo(boolean type) has been changed to shc_method(character type). From now, shc_method is only qiao or ryoo in order to apply berry_task = shc.

2. A test case is included.

3. The user guide have been updated. Although it explains, the difference between the two methods is the degree of approximation when obtaining spin-current matrix from Wannier interpolation.

4. I removed debug lines commented out.

5. I completed commit to pw2wannier90.x: https://gitlab.com/QEF/q-e/-/merge_requests/1249

Regards,
Minsu Ghim
Seoul National University

### CODE REVIEW

File: test-suite/tests/testpostw90_fe_ahc_adaptandfermi/Fe.wpout [hunk 1]
```diff
@@ -0,0 +1,315 @@
+
```
Comment (jaemolihm) [inline test-suite/tests/testpostw90_fe_ahc_adaptandfermi/Fe.wpout]: As far as I understand, this file, as well as `test-suite/tests/testpostw90_fe_ahc_adaptandfermi/Fe.mmn` should not be included. Especially, `Fe.mmn` is a large file that is intended to exist only locally.
(It seems they are included by the `git add .` command, due to a missing `.gitignore` in this folder: see #352.)

Just deleting them in another commit may *not* be sufficient as these files will remain in the history, increasing the size of the repository, but I am not sure.

File: src/postw90/berry.F90 [hunk 2]
```diff
@@ -1992,43 +2014,82 @@ subroutine berry_get_js_k(kpt, eig, del_alpha_eig, D_alpha_h, UU, js_k)
...
+        ! to get spin current, we need to multiply it by hbar/2,
+        ! also we need to divide it by hbar to recover the velocity
+        ! operator, these are done outside of this subroutine
+        js_k = 1.0_dp/2.0_dp*(B_k + conjg(transpose(B_k)))
+
+      else !if Ryoo (PRB RPS19 Eq.(21))
+        do j = shc_alpha, shc_alpha
+          do i = shc_gamma, shc_gamma !RPS19 Eqs.(37)-(40)
+            call pw90common_fourier_R_to_k_new(kpt, SAA_R(:, :, :, i, j), OO=SAA(:, :, i, j))
+            call pw90common_fourier_R_to_k_new(kpt, SBB_R(:, :, :, i, j), OO=SBB(:, :, i, j))
```
Comment (jaemolihm) [inline src/postw90/berry.F90]: Maybe you can just use `call pw90common_fourier_R_to_k_new(kpt, SBB_R(:, :, :, shc_gamma, shc_alpha), OO=SBB(:, :, shc_gamma, shc_alpha))`?

File: src/postw90/berry.F90 [hunk 3]
```diff
@@ -1992,43 +2014,82 @@ subroutine berry_get_js_k(kpt, eig, del_alpha_eig, D_alpha_h, UU, js_k)
...
+                                           OO_dx=delHH(:, :, 1), &
+                                           OO_dy=delHH(:, :, 2), &
+                                           OO_dz=delHH(:, :, 3))
+
+        VV0(:, :) = utility_rotate(delHH_alpha(:, :), UU, num_wann)
+        !if(.true.) write(*,*) "kpt",kpt,"SA_K",real(SAA(1,:,1,3),dp)
+        do i = 1, 3
+          do j = 1, 3
+            SAA(:, :, i, j) = utility_rotate(SAA(:, :, i, j), UU, num_wann)
+            SBB(:, :, i, j) = utility_rotate(SBB(:, :, i, j), UU, num_wann)
```
Comment (jaemolihm) [inline src/postw90/berry.F90]: This part seems to be inconsistent with the `call pw90common_fourier_R_to_k_new` part, where only `SBB(:, :, shc_gamma, shc_alpha)` is calculated.

File: src/postw90/berry.F90 [hunk 4]
```diff
@@ -1992,43 +2014,82 @@ subroutine berry_get_js_k(kpt, eig, del_alpha_eig, D_alpha_h, UU, js_k)
...
+        do n = 1, num_wann
+          do m = 1, num_wann !RPS19 Eq.(21) and Eq.(26)
+            js_k(n, m) = spinVel0(n, m) &
+                         - cmplx_i*(eig(m)*SAA(n, m, shc_gamma, shc_alpha) - SBB(n, m, shc_gamma, shc_alpha))
+            js_k(n, m) = js_k(n, m) &
+                         + cmplx_i*(eig(n)*conjg(SAA(m, n, shc_gamma, shc_alpha)) - conjg(SBB(m, n, shc_gamma, shc_alpha)))
+          enddo
+        enddo
+        js_k = js_k/2.0_dp
+        !js_k=1.0_dp
```
Comment (jaemolihm) [inline src/postw90/berry.F90]: If this line is only for debugging, I think you can just delete it.

File: src/postw90/get_oper.F90 [hunk 5]
```diff
@@ -55,6 +55,15 @@ module w90_get_oper
   complex(kind=dp), allocatable, save :: SH_R(:, :, :, :) ! <0n|sigma_x,y,z.H|Rm>
   !! $$\langle 0n | \sigma_{x,y,z}.H  | Rm \rangle$$
 
+  !temporary
+  ! <0n|sigma_a. (r-R)_b|Rm>
+  !
+  complex(kind=dp), allocatable, save :: SAA_R(:, :, :, :, :)
```
Comment (jaemolihm) [inline src/postw90/get_oper.F90]: Change `!temporary` to e.g. `!Spin Hall using Ryoo method`? Also, the comments could be made more consistent with the previous lines.

File: src/postw90/get_oper.F90 [hunk 6]
```diff
@@ -1444,6 +1458,281 @@ subroutine get_SHC_R
...
+    complex(kind=dp), allocatable :: H_q_qb2(:, :)
+    real(kind=dp)                 :: c_real, c_img
+    character(len=60)             :: header
+
+    if (timing_level > 1 .and. on_root) call io_stopwatch('get_oper: get_CC_R', 1)
+
+    if (.not. allocated(BB_R)) then
+      allocate (SBB_R(num_wann, num_wann, nrpts, 3, 3))
+    else
+      if (timing_level > 1 .and. on_root) call io_stopwatch('get_oper: get_CC_R', 2)
```
Comment (jaemolihm) [inline src/postw90/get_oper.F90]: `get_CC_R` -> `get_SBB_R`, same in other `io_stopwatch` calls in `get_SBB_R` and `get_SAA_R`.

File: src/postw90/get_oper.F90 [hunk 7]
```diff
@@ -1444,6 +1458,281 @@ subroutine get_SHC_R
...
+          qb2 = nnlist(ik, nn2)
+          call get_win_min(qb2, winmin_qb2)
+          do ipol = 1, 3
+            !
+            ! Read from .sIu file the matrices <u_q|s_a|u_{q+b2}>
+            ! between the original ab initio eigenstates
+            !
+            read (sIu_in, err=114, end=114) &
+              ((Ho_q_qb2(n, m, ipol), n=1, num_bands), m=1, num_bands)
+            ! pw2wannier90 is coded a bit strangely, so here we take the transpose
```
Comment (jaemolihm) [inline src/postw90/get_oper.F90:1685]: An explanation on how the matrix elements calculated by pw2wannier90 differ from what we need here would be helpful.

**Comment by manxkim** [comment]
I discovered that the explanation of the difference between the two method is wrong. I will soon make it correct, but in a nutshell, Ryoo's method does not generally use the identity matrix to find spin-current matrix, but another derivtaion.

File: src/postw90/berry.F90 (hunk 2 above)
Comment (manxkim) [reply src/postw90/berry.F90]: That will be better. These do sentences are the trace of considering multiple shc_alpha and shc_gamma, but it is now obsolete.

File: src/postw90/berry.F90 (hunk 3 above)
Comment (manxkim) [reply src/postw90/berry.F90]: modified as you mentioned.

File: src/postw90/get_oper.F90 (hunk 7 above)
Comment (manxkim) [reply src/postw90/get_oper.F90:1685]: Simple, as it says, pw2wannier90 is coded with an unnecessary transposition.

**Review by ivosouza0 — APPROVED** [review]
All looks good!

**Pull request merged**